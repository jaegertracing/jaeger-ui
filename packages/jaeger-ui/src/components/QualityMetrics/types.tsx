// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';

import { TExample } from '../common/ExamplesLink';
import { TColumnDef, TRow } from '../common/DetailsCard/types';

export type TQualityMetrics = {
  traceQualityDocumentationLink: string;
  bannerText?:
    | string
    | {
        value: string;
        styling: React.CSSProperties;
      };
  scores: {
    key: string;
    label: string;
    max: number;
    value: number;
  }[];
  metrics: {
    name: string;
    category: string; // matching some scores.key;
    description: string;
    metricDocumentationLink: string;
    metricWeight: number;
    passCount: number;
    passExamples?: TExample[];
    failureCount: number;
    failureExamples?: TExample[];
    exemptionCount?: number;
    exemptionExamples?: TExample[];
    details?: {
      description?: string;
      header?: string;
      columns: TColumnDef[];
      rows: TRow[];
    }[];
  }[];
  clients?: {
    version: string;
    minVersion: string;
    count: number;
    examples: TExample[];
  }[];
};
