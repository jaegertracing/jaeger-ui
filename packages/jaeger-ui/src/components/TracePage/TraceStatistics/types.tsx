// Copyright (c) 2020 The Jaeger Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

export interface ITableSpan {
  type: string;
  name: string;
  count: number;
  total: number;
  avg: number;
  min: number;
  max: number;
  selfTotal: number;
  selfAvg: number;
  selfMin: number;
  selfMax: number;
  percent: number;
  isDetail: boolean;
  parentElement: string;
  color: string; // If it is a service name, the color will be set.
  searchColor: string;
  colorToPercent: string; // Color created by percent
}

export interface ITableValues {
  uid: string;
  value: number;
}

export interface IColumnValue {
  title: string;
  attribute: string;
  suffix: string;
  isDecimal: boolean;
}
