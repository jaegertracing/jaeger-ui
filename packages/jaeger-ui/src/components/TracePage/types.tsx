// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { TNil } from '../../types';

interface ITimeCursorUpdate {
  cursor: number | TNil;
}

interface ITimeReframeUpdate {
  reframe: {
    anchor: number;
    shift: number;
  };
}

interface ITimeShiftEndUpdate {
  shiftEnd: number;
}

interface ITimeShiftStartUpdate {
  shiftStart: number;
}

export type TUpdateViewRangeTimeFunction = (start: number, end: number, trackSrc?: string) => void;

export type ViewRangeTimeUpdate =
  | ITimeCursorUpdate
  | ITimeReframeUpdate
  | ITimeShiftEndUpdate
  | ITimeShiftStartUpdate;

export interface IViewRangeTime {
  current: [number, number];
  cursor?: number | TNil;
  reframe?: {
    anchor: number;
    shift: number;
  };
  shiftEnd?: number;
  shiftStart?: number;
}

export interface IViewRange {
  time: IViewRangeTime;
}
