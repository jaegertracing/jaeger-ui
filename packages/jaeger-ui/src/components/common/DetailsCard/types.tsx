// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import type React from 'react';
import { TExample } from '../ExamplesLink';

export type TStyledValue = {
  linkTo?: string;
  styling?: React.CSSProperties;
  value: string | React.ReactElement;
};

export type TColumnDef = {
  key: string;
  label?: string;
  preventSort?: boolean;
  styling?: React.CSSProperties;
};

export type TColumnDefs = (string | TColumnDef)[];

export type TRow = Record<string, TStyledValue | string | number | TExample[]>;

export type TDetails = string | string[] | TRow[];

export type TFilterDropdownProps = {
  clearFilters?: () => void;
  confirm: () => void;
  selectedKeys: React.Key[];
  setSelectedKeys: (selectedKeys: React.Key[]) => void;
};
