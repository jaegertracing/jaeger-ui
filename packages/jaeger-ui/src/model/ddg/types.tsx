// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { TVertex } from '@jaegertracing/plexus/lib/types';

import PathElem from './PathElem';

export { default as PathElem } from './PathElem';

export enum EViewModifier {
  None,
  Hovered,
  Selected,
  Emphasized = 1 << 2, // eslint-disable-line no-bitwise
  PathHovered = 1 << 3, // eslint-disable-line no-bitwise
}

export enum EDdgDensity {
  ExternalVsInternal = 'ext-vs-int',
  MostConcise = 'mc',
  OnePerLevel = 'per-level',
  PreventPathEntanglement = 'ppe',
  UpstreamVsDownstream = 'up-vs-down',
}

export enum ECheckedStatus {
  Empty = 'Empty',
  Full = 'Full',
  Partial = 'Partial',
}

export enum EDirection {
  Upstream = -1,
  Downstream = 1,
}

export type TDdgPayloadEntry = {
  operation: string;
  service: string;
};

export type TDdgPayloadPath = {
  path: TDdgPayloadEntry[];
  // TODO: Everett Tech Debt: Fix KeyValuePair types
  attributes: {
    key: 'exemplar_trace_id'; // eslint-disable-line camelcase
    value: string;
  }[];
};

export type TDdgPayload = {
  dependencies: TDdgPayloadPath[];
};

export type TDdgService = {
  name: string;
  operations: Map<string, TDdgOperation>;
};

export type TDdgOperation = {
  name: string;
  pathElems: PathElem[];
  service: TDdgService;
};

export type TDdgServiceMap = Map<string, TDdgService>;

export type TDdgPath = {
  focalIdx: number;
  members: PathElem[];
  traceIDs: string[];
};

export type TDdgDistanceToPathElems = Map<number, PathElem[]>;

export type TDdgModel = {
  distanceToPathElems: TDdgDistanceToPathElems;
  hash: string;
  paths: TDdgPath[];
  services: TDdgServiceMap;
  visIdxToPathElem: PathElem[];
};

export type TDdgVertex = TVertex<{
  isFocalNode: boolean;
  key: string;
  operation: string | string[] | null;
  service: string;
}>;

export type TDdgSparseUrlState = {
  density: EDdgDensity;
  decoration?: string;
  end?: number;
  hash?: string;
  operation?: string;
  service?: string;
  showOp?: boolean;
  start?: number;
  visEncoding?: string;
};

export type TDdgModelParams = {
  service: string;
  operation?: string;
  start: number;
  end: number;
};

export type TDdgActionMeta = {
  query: TDdgModelParams;
};

export type TDdgAddViewModifierPayload = TDdgModelParams & {
  // Number instead of EViewModifier so that multiple views can be changed at once.
  viewModifier: number;
  visibilityIndices: number[];
};

export type TDdgClearViewModifiersFromIndicesPayload = TDdgAddViewModifierPayload & { viewModifier?: void };

export type TDdgRemoveViewModifierFromIndicesPayload = TDdgAddViewModifierPayload;

export type TDdgRemoveViewModifierPayload = TDdgAddViewModifierPayload & { visibilityIndices?: void };

export type TDdgViewModifierRemovalPayload =
  | TDdgClearViewModifiersFromIndicesPayload
  | TDdgRemoveViewModifierFromIndicesPayload
  | TDdgRemoveViewModifierPayload;

export type THop = { distance: number; fullness: ECheckedStatus };
