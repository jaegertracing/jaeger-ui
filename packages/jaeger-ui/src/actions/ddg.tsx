// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import _identity from 'lodash/identity';
import { createActions, ActionFunctionAny, Action } from 'redux-actions';

import {
  TDdgAddViewModifierPayload,
  TDdgClearViewModifiersFromIndicesPayload,
  TDdgRemoveViewModifierFromIndicesPayload,
  TDdgRemoveViewModifierPayload,
  TDdgViewModifierRemovalPayload,
} from '../model/ddg/types';
import generateActionTypes from '../utils/generate-action-types';

export const actionTypes = generateActionTypes('@jaeger-ui/DEEP-DEPENDENCY-GRAPH', [
  'ADD_VIEW_MODIFIER',
  'CLEAR_VIEW_MODIFIERS_FROM_INDICES',
  'REMOVE_VIEW_MODIFIER',
  'REMOVE_VIEW_MODIFIER_FROM_INDICES',
]);

const addViewModifier: (kwarg: TDdgAddViewModifierPayload) => TDdgAddViewModifierPayload = _identity;
const clearViewModifiersFromIndices: (
  kwarg: TDdgClearViewModifiersFromIndicesPayload
) => TDdgClearViewModifiersFromIndicesPayload = _identity;
const removeViewModifier: (kwarg: TDdgRemoveViewModifierPayload) => TDdgRemoveViewModifierPayload = _identity;
const removeViewModifierFromIndices: (
  kwarg: TDdgRemoveViewModifierFromIndicesPayload
) => TDdgRemoveViewModifierFromIndicesPayload = _identity;

const fullActions = createActions<TDdgAddViewModifierPayload | TDdgViewModifierRemovalPayload>({
  [actionTypes.ADD_VIEW_MODIFIER]: addViewModifier,
  [actionTypes.CLEAR_VIEW_MODIFIERS_FROM_INDICES]: clearViewModifiersFromIndices,
  [actionTypes.REMOVE_VIEW_MODIFIER]: removeViewModifier,
  [actionTypes.REMOVE_VIEW_MODIFIER_FROM_INDICES]: removeViewModifierFromIndices,
});

type FullActionsType = {
  jaegerUi: {
    deepDependencyGraph: Record<string, ActionFunctionAny<Action<TDdgViewModifierRemovalPayload>>>;
  };
};

export default (fullActions as unknown as FullActionsType).jaegerUi.deepDependencyGraph;
