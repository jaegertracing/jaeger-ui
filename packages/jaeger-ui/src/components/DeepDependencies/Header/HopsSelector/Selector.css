/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.HopsSelector--Selector {
  color: var(--tx-color-muted);
  margin: 0 0.5em;
  padding: 0px 3px;
}

.HopsSelector--Selector--decrement,
.HopsSelector--Selector--increment {
  background-color: #f4f4f4;
  border: none;
  cursor: pointer;
  height: 27px;
  padding: 0em 0.3em 0.3em;
  width: 27px;
}

.HopsSelector--Selector--decrement {
  margin-right: 0.75em;
}

.HopsSelector--Selector--increment {
  margin-left: 0.75em;
}

.HopsSelector--Selector--decrement:disabled,
.HopsSelector--Selector--increment:disabled {
  color: #bbb;
}

.HopsSelector--Selector--btn {
  background: transparent;
  border: none;
  color: #096dd9;
  cursor: pointer;
  font-weight: bold;
  margin: 0 0.2em;
  padding: 1px 2px;
}

.HopsSelector--Selector--btn.is-Empty {
  color: #aaa;
  font-weight: 400;
}

.HopsSelector--Selector--btn.is-Partial {
  color: #d46b08;
}

/* .HopsSelector--Selector > .HopsSelector--Selector--furthest {
  margin-left: 0.3em;
} */

.HopsSelector--Selector > .HopsSelector--Selector--delimiter {
  color: #999;
  font-weight: 400;
}

.HopsSelector--Selector--slash {
  color: #999;
}

.HopsSelector--Selector--AscIcon {
  margin-right: 0.3em;
}

.HopsSelector--Selector--AscIcon.is-Up {
  transform: scaleY(-1);
}

.HopsSelector--Selector--ChevronDown {
  color: #999;
  height: 0.7em;
  padding-bottom: 0.15em;
  width: 0.7em;
}

.HopsSelector--Selector--ChevronRight {
  color: #999;
  height: 0.7em;
  padding-bottom: 0.15em;
  width: 0.7em;
}

.HopsSelector--Selector--ChevronRight.is-Empty {
  color: #ccc;
}
