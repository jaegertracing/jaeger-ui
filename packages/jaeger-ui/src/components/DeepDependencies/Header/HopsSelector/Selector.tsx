// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { PureComponent } from 'react';
import { Popover } from 'antd';
import { ImSortAmountAsc } from 'react-icons/im';
import { FaChevronRight } from 'react-icons/fa';

import ChevronDown from '../ChevronDown';
import { trackHopChange } from '../../index.track';
import { ECheckedStatus, EDirection, THop } from '../../../../model/ddg/types';

import './Selector.css';

type TProps = {
  direction: EDirection;
  furthestDistance: number;
  furthestFullDistance: number;
  furthestFullness: ECheckedStatus;
  handleClick: (distance: number, direction: EDirection) => void;
  hops: THop[];
};

const CLASSNAME = 'HopsSelector--Selector';

export default class Selector extends PureComponent<TProps> {
  private handleClick(distance: number) {
    const { direction, furthestFullDistance, handleClick } = this.props;
    handleClick(distance, direction);
    trackHopChange(furthestFullDistance, distance, direction);
  }

  private makeBtn = (
    { distance, fullness, suffix = 'popover-content' }: THop & { suffix?: string },
    showChevron?: number
  ) => {
    const { direction } = this.props;
    return (
      <React.Fragment key={`${distance} ${direction} ${suffix}`}>
        {Boolean(showChevron) && <FaChevronRight className={`${CLASSNAME}--ChevronRight is-${fullness}`} />}
        <button
          className={`${CLASSNAME}--btn is-${fullness} ${CLASSNAME}--${suffix}`}
          type="button"
          onClick={() => this.handleClick(distance)}
        >
          {Math.abs(distance)}
        </button>
      </React.Fragment>
    );
  };

  render() {
    const { direction, furthestDistance, furthestFullness, handleClick, hops } = this.props;
    const streamText = direction === EDirection.Downstream ? 'Down' : 'Up';
    const streamLabel = `${streamText}stream hops`;
    const lowercaseLabel = streamLabel.toLowerCase();
    if (hops.length <= 1) {
      return <span className={CLASSNAME}>No {lowercaseLabel}</span>;
    }

    const decrementBtn = (
      <button
        key={`decrement ${direction}`}
        disabled={furthestDistance === 0}
        className={`${CLASSNAME}--decrement`}
        type="button"
        onClick={() => handleClick(furthestDistance - direction, direction)}
      >
        -
      </button>
    );
    const incrementBtn = (
      <button
        key={`increment ${direction}`}
        disabled={furthestDistance === hops[hops.length - 1].distance}
        className={`${CLASSNAME}--increment`}
        type="button"
        onClick={() => handleClick(furthestDistance + direction, direction)}
      >
        +
      </button>
    );
    const delimiterBtn = this.makeBtn({
      ...hops[hops.length - 1],
      suffix: 'delimiter',
    });
    const furthestBtn = this.makeBtn({
      distance: furthestDistance,
      fullness: furthestFullness,
      suffix: 'furthest',
    });

    return (
      <Popover
        arrowPointAtCenter
        content={[decrementBtn, ...hops.map(this.makeBtn), incrementBtn]}
        placement="bottom"
        title={`Visible ${lowercaseLabel}`}
      >
        <span className={CLASSNAME}>
          <ImSortAmountAsc className={`${CLASSNAME}--AscIcon is-${streamText}`} />
          {streamLabel}
          {furthestBtn}
          <span className={`${CLASSNAME}--delimiter`}>/</span>
          {delimiterBtn}
          {/* <IoChevronDown className={`${CLASSNAME}--ChevronDown`} /> */}
          <ChevronDown className="ub-ml1" />
        </span>
      </Popover>
    );
  }
}
