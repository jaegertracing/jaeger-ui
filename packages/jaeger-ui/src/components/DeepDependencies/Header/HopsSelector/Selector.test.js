// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { shallow } from 'enzyme';
import { Popover } from 'antd';

import { EDirection } from '../../../../model/ddg/types';
import Selector from './Selector';

describe('Selector', () => {
  const handleClick = jest.fn();
  const hops = ['full', 'partial', 'full', 'empty', 'empty'];
  const makeHops = (fullnessArr, direction = EDirection.Downstream) =>
    fullnessArr.map((fullness, i) => ({
      distance: i * direction,
      fullness,
    }));
  const props = {
    furthestDistance: 2,
    furthestFullness: 'full',
    hops: makeHops(hops),
    direction: EDirection.Downstream,
    handleClick,
  };
  let wrapper;
  function getPopoverButtons() {
    const popoverContent = wrapper.find(Popover).prop('content');
    return shallow(<div>{popoverContent}</div>).find('button');
  }

  beforeEach(() => {
    wrapper = shallow(<Selector {...props} />);
  });

  it('renders message when there are no options', () => {
    const message = shallow(<Selector hops={makeHops(['full'])} />);
    expect(message).toMatchSnapshot();
  });

  it('renders buttons with expected text and classNames', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders upstream hops with negative distance correctly', () => {
    const upstreamProps = {
      direction: EDirection.Upstream,
      furthestDistance: EDirection.Upstream * props.furthestDistance,
      furthestFullness: 'full',
      hops: makeHops(hops, EDirection.Upstream),
      handleClick,
    };
    const upstreamWrapper = shallow(<Selector {...upstreamProps} />);
    expect(upstreamWrapper).toMatchSnapshot();
  });

  it('calls handleClick with correct arguments from label buttons', () => {
    const buttons = wrapper.find('button');
    expect(buttons.length).toBe(2);

    buttons.first().simulate('click');
    expect(handleClick).toHaveBeenLastCalledWith(props.furthestDistance, props.direction);

    buttons.last().simulate('click');
    expect(handleClick).toHaveBeenLastCalledWith(props.hops.length - 1, props.direction);
  });

  it('calls handleClick with correct arguments from popover buttons', () => {
    const expectedDistances = [1, 0, 1, 2, 3, 4, 3];
    const popoverButtons = getPopoverButtons();
    expect(popoverButtons.length).toBe(expectedDistances.length);

    popoverButtons.forEach((btn, i) => {
      btn.simulate('click');
      expect(handleClick).toHaveBeenLastCalledWith(expectedDistances[i], props.direction);
    });
  });

  it('disables increment/decrement buttons', () => {
    wrapper.setProps({ furthestDistance: 0 });
    expect(
      getPopoverButtons()
        .first()
        .prop('disabled')
    ).toBe(true);

    wrapper.setProps({ furthestDistance: hops.length - 1 });
    expect(
      getPopoverButtons()
        .last()
        .prop('disabled')
    ).toBe(true);
  });
});
