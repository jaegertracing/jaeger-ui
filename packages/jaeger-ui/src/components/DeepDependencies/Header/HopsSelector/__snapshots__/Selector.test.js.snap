// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selector renders buttons with expected text and classNames 1`] = `
<Popover
  arrowPointAtCenter={true}
  content={
    Array [
      <button
        className="HopsSelector--Selector--decrement"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        -
      </button>,
      <React.Fragment>
        <button
          className="HopsSelector--Selector--btn is-full HopsSelector--Selector--popover-content"
          onClick={[Function]}
          type="button"
        >
          0
        </button>
      </React.Fragment>,
      <React.Fragment>
        <IoChevronRight
          className="HopsSelector--Selector--ChevronRight is-partial"
        />
        <button
          className="HopsSelector--Selector--btn is-partial HopsSelector--Selector--popover-content"
          onClick={[Function]}
          type="button"
        >
          1
        </button>
      </React.Fragment>,
      <React.Fragment>
        <IoChevronRight
          className="HopsSelector--Selector--ChevronRight is-full"
        />
        <button
          className="HopsSelector--Selector--btn is-full HopsSelector--Selector--popover-content"
          onClick={[Function]}
          type="button"
        >
          2
        </button>
      </React.Fragment>,
      <React.Fragment>
        <IoChevronRight
          className="HopsSelector--Selector--ChevronRight is-empty"
        />
        <button
          className="HopsSelector--Selector--btn is-empty HopsSelector--Selector--popover-content"
          onClick={[Function]}
          type="button"
        >
          3
        </button>
      </React.Fragment>,
      <React.Fragment>
        <IoChevronRight
          className="HopsSelector--Selector--ChevronRight is-empty"
        />
        <button
          className="HopsSelector--Selector--btn is-empty HopsSelector--Selector--popover-content"
          onClick={[Function]}
          type="button"
        >
          4
        </button>
      </React.Fragment>,
      <button
        className="HopsSelector--Selector--increment"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        +
      </button>,
    ]
  }
  mouseEnterDelay={0.1}
  mouseLeaveDelay={0.1}
  overlayStyle={Object {}}
  placement="bottom"
  title="Visible downstream hops"
  transitionName="zoom-big"
  trigger="hover"
>
  <span
    className="HopsSelector--Selector"
  >
    <FaSortAmountAsc
      className="HopsSelector--Selector--AscIcon is-Down"
    />
    Downstream hops
    <button
      className="HopsSelector--Selector--btn is-full HopsSelector--Selector--furthest"
      onClick={[Function]}
      type="button"
    >
      2
    </button>
    <span
      className="HopsSelector--Selector--delimiter"
    >
      /
    </span>
    <button
      className="HopsSelector--Selector--btn is-empty HopsSelector--Selector--delimiter"
      onClick={[Function]}
      type="button"
    >
      4
    </button>
    <Memo(ChevronDown)
      className="ub-ml1"
    />
  </span>
</Popover>
`;

exports[`Selector renders message when there are no options 1`] = `
<span
  className="HopsSelector--Selector"
>
  No 
  upstream hops
</span>
`;

exports[`Selector renders upstream hops with negative distance correctly 1`] = `
<Popover
  arrowPointAtCenter={true}
  content={
    Array [
      <button
        className="HopsSelector--Selector--decrement"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        -
      </button>,
      <React.Fragment>
        <button
          className="HopsSelector--Selector--btn is-full HopsSelector--Selector--popover-content"
          onClick={[Function]}
          type="button"
        >
          0
        </button>
      </React.Fragment>,
      <React.Fragment>
        <IoChevronRight
          className="HopsSelector--Selector--ChevronRight is-partial"
        />
        <button
          className="HopsSelector--Selector--btn is-partial HopsSelector--Selector--popover-content"
          onClick={[Function]}
          type="button"
        >
          1
        </button>
      </React.Fragment>,
      <React.Fragment>
        <IoChevronRight
          className="HopsSelector--Selector--ChevronRight is-full"
        />
        <button
          className="HopsSelector--Selector--btn is-full HopsSelector--Selector--popover-content"
          onClick={[Function]}
          type="button"
        >
          2
        </button>
      </React.Fragment>,
      <React.Fragment>
        <IoChevronRight
          className="HopsSelector--Selector--ChevronRight is-empty"
        />
        <button
          className="HopsSelector--Selector--btn is-empty HopsSelector--Selector--popover-content"
          onClick={[Function]}
          type="button"
        >
          3
        </button>
      </React.Fragment>,
      <React.Fragment>
        <IoChevronRight
          className="HopsSelector--Selector--ChevronRight is-empty"
        />
        <button
          className="HopsSelector--Selector--btn is-empty HopsSelector--Selector--popover-content"
          onClick={[Function]}
          type="button"
        >
          4
        </button>
      </React.Fragment>,
      <button
        className="HopsSelector--Selector--increment"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        +
      </button>,
    ]
  }
  mouseEnterDelay={0.1}
  mouseLeaveDelay={0.1}
  overlayStyle={Object {}}
  placement="bottom"
  title="Visible upstream hops"
  transitionName="zoom-big"
  trigger="hover"
>
  <span
    className="HopsSelector--Selector"
  >
    <FaSortAmountAsc
      className="HopsSelector--Selector--AscIcon is-Up"
    />
    Upstream hops
    <button
      className="HopsSelector--Selector--btn is-full HopsSelector--Selector--furthest"
      onClick={[Function]}
      type="button"
    >
      2
    </button>
    <span
      className="HopsSelector--Selector--delimiter"
    >
      /
    </span>
    <button
      className="HopsSelector--Selector--btn is-empty HopsSelector--Selector--delimiter"
      onClick={[Function]}
      type="button"
    >
      4
    </button>
    <Memo(ChevronDown)
      className="ub-ml1"
    />
  </span>
</Popover>
`;
